Delete OperatorOptions, absorb AliasAnalysisKind into FunctionSchema.

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #34456 WIP
* #34455 Add torch::jit::parseName (mirroring parseSchema/parseSchemaOrName)
* #34221 Relax internal invariants of Dispatcher in preparation for schema-less registration
* **#34160 Delete OperatorOptions, absorb AliasAnalysisKind into FunctionSchema.**
* #33838 Beef up documentation on Dispatcher.h, reorder methods for clarity.

I constructed the patch by deleting OperatorOptions and then rerouting
all queries for AliasAnalysisKind to FunctionSchema.  Some of the
behavior is kind of bogus: we really shouldn't be mutating FunctionSchema
after the fact, but that won't get fixed until we actually switch to
true schema merging.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20282846](https://our.internmc.facebook.com/intern/diff/D20282846)

ezyang committed:
Delete OperatorOptions, absorb AliasAnalysisKind into FunctionSchema.

I constructed the patch by deleting OperatorOptions and then rerouting
all queries for AliasAnalysisKind to FunctionSchema.  Some of the
behavior is kind of bogus: we really shouldn't be mutating FunctionSchema
after the fact, but that won't get fixed until we actually switch to
true schema merging.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Delete OperatorOptions, absorb AliasAnalysisKind into FunctionSchema."


I constructed the patch by deleting OperatorOptions and then rerouting
all queries for AliasAnalysisKind to FunctionSchema.  Some of the
behavior is kind of bogus: we really shouldn't be mutating FunctionSchema
after the fact, but that won't get fixed until we actually switch to
true schema merging.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Delete OperatorOptions, absorb AliasAnalysisKind into FunctionSchema."


I constructed the patch by deleting OperatorOptions and then rerouting
all queries for AliasAnalysisKind to FunctionSchema.  Some of the
behavior is kind of bogus: we really shouldn't be mutating FunctionSchema
after the fact, but that won't get fixed until we actually switch to
true schema merging.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Delete OperatorOptions, absorb AliasAnalysisKind into FunctionSchema."


I constructed the patch by deleting OperatorOptions and then rerouting
all queries for AliasAnalysisKind to FunctionSchema.  Some of the
behavior is kind of bogus: we really shouldn't be mutating FunctionSchema
after the fact, but that won't get fixed until we actually switch to
true schema merging.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Delete OperatorOptions, absorb AliasAnalysisKind into FunctionSchema."


I constructed the patch by deleting OperatorOptions and then rerouting
all queries for AliasAnalysisKind to FunctionSchema.  Some of the
behavior is kind of bogus: we really shouldn't be mutating FunctionSchema
after the fact, but that won't get fixed until we actually switch to
true schema merging.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

smessmer commented on aten/src/ATen/core/function_schema.h:
```diff
@@ -188,6 +190,13 @@ struct FunctionSchema {
+  // if no alias information is directly specified, what kind of "default"
+  // alias information should we infer?
+  // NB: due to alias analysis kind merging, this may be nullopt.  Eventually
+  // this should always be set no matter what
```
Eventually, this should not be "always set" but I would argue it should go away entirely since this information is already fully representable in the `alias_info_` member. No action on this PR though.

ezyang replied:
Yeah I think I agree with this.

smessmer commented on test/cpp/jit/test_alias_analysis.cpp:
```diff
@@ -9,10 +9,8 @@
-  c10::OperatorOptions result;
-  result.setAliasAnalysis(c10::AliasAnalysisKind::FROM_SCHEMA);
-  return result;
+inline c10::AliasAnalysisKind aliasAnalysisFromSchema() {
```
nit: doensn't look like we'd need to factor this out into a separate function anymore. Same for several similar functions below in this changeset.

ezyang replied:
Yeah, I saw this too, chose not to inline to reduce footprint of this patch

smessmer commented on torch/csrc/jit/runtime/operator.h:
```diff
@@ -66,26 +66,25 @@ struct TORCH_API Operator {
   Operator(c10::OperatorHandle opHandle, Operation operation)
       : schema_(std::make_shared<FunctionSchema>(opHandle.schema())),
         op_(std::make_shared<Operation>(std::move(operation))),
-        c10Handle_(opHandle),
```
note: any changes in operator.h are likely to collide with https://github.com/pytorch/pytorch/pull/33905

ezyang replied:
I took a look, the merge conflict looks easy to resolve.

ezyang committed:
Update on "Delete OperatorOptions, absorb AliasAnalysisKind into FunctionSchema."


I constructed the patch by deleting OperatorOptions and then rerouting
all queries for AliasAnalysisKind to FunctionSchema.  Some of the
behavior is kind of bogus: we really shouldn't be mutating FunctionSchema
after the fact, but that won't get fixed until we actually switch to
true schema merging.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Delete OperatorOptions, absorb AliasAnalysisKind into FunctionSchema."


I constructed the patch by deleting OperatorOptions and then rerouting
all queries for AliasAnalysisKind to FunctionSchema.  Some of the
behavior is kind of bogus: we really shouldn't be mutating FunctionSchema
after the fact, but that won't get fixed until we actually switch to
true schema merging.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Delete OperatorOptions, absorb AliasAnalysisKind into FunctionSchema."


I constructed the patch by deleting OperatorOptions and then rerouting
all queries for AliasAnalysisKind to FunctionSchema.  Some of the
behavior is kind of bogus: we really shouldn't be mutating FunctionSchema
after the fact, but that won't get fixed until we actually switch to
true schema merging.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Delete OperatorOptions, absorb AliasAnalysisKind into FunctionSchema."


I constructed the patch by deleting OperatorOptions and then rerouting
all queries for AliasAnalysisKind to FunctionSchema.  Some of the
behavior is kind of bogus: we really shouldn't be mutating FunctionSchema
after the fact, but that won't get fixed until we actually switch to
true schema merging.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

bhosmer reviewed (approved):
This seems fine to me. But what you're having to do here re-raises the idea of eliminating the redundancy between c10 and JIT registries. 

Aside: if we partition the registries so JIT is only non-dispatched primitives and c10 is all dispatched ops, then the JIT registry doesn't need dynamic capabilities at all, right?

bhosmer commented:
Longer term, if they both still exist, it'll be nice if the two registries have consistent APIs where they overlap (e.g. same way of specifying schemas, alias analysis), as opposed to the current situation where c10 has the kwarg-style `Options` layer but JIT is just simple positional calls.

ezyang committed:
Update on "Delete OperatorOptions, absorb AliasAnalysisKind into FunctionSchema."


I constructed the patch by deleting OperatorOptions and then rerouting
all queries for AliasAnalysisKind to FunctionSchema.  Some of the
behavior is kind of bogus: we really shouldn't be mutating FunctionSchema
after the fact, but that won't get fixed until we actually switch to
true schema merging.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20282846](https://our.internmc.facebook.com/intern/diff/D20282846)

[ghstack-poisoned]
